### ISSUE TITLE
TRAPI-1.3-compliant node expansion

### ISSUE THREAD

**Issue by colleenXu** [biothings/biothings_explorer#497]
Related to #366 and #196 and these PRs: https://github.com/biothings/bte-trapi-workspace/pull/17 and https://github.com/biothings/bte_trapi_query_graph_handler/pull/113

I looked at the current behavior of the linked PRs and made a quick [ask](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1235002384) for clarification. 

However, I think that the requirements around "query node ID expansion" need to be articulated, given the changes in TRAPI 1.3 and other changes since the previous linked issues were opened. 

---

## A. Creative-mode

Translator's standards are that only 1 input ID for a Disease is expected for creative-mode, and BTE checks that the QGraph meets this standard for creative-mode. However, the current PRs seem to expand the Disease QNode ID (from 1 ID to it and its descendants) before this check....so BTE will stop execution and return an empty KG / results section and a TRAPI log that says:

```
        {
            "timestamp": "2022-09-02T01:40:50.275Z",
            "level": "WARNING",
            "message": "Inferred Mode queries with multiple IDs are not supported. Your query terminates.",
            "code": null
        }
```

As for what to do...I'm thinking we still want QNode ID expansion to happen, after the the checks for creative-mode to happen on the original QGraph. Then, some [templates](https://github.com/biothings/bte_trapi_query_graph_handler/tree/main/data/templates/Drug-treats-Disease) will need to be adjusted to remove the `subclass_of` hops. 

EDIT: see below, we chose to turn off ID-expansion for creative-mode instead. This means more flexibility in how creative-mode queries work (you can choose to "do ID-expansion" through subclass_of QEdges, or choose not to have them).

---

## B. Manipulation of QGraph and the results section format

TRAPI 1.3 introduces a NodeBinding.query_id property, and I interpret TRAPI 1.3 in the following way:
* don't manipulate the QNode with the original ID(s) by adding the descendant IDs
* if QNode ID expansion was done and a result uses that QNode's descendant ID entity, in that result's NodeBinding:
    * the `id` is the descendant ID entity's primary ID
    * the `query_id` is the original QNode ID 

<details>
<summary>example of what the PRs currently do vs what I outlined above</summary>

This query has a QNode with the [MONDO ID for noonan syndrome](https://monarchinitiative.org/disease/MONDO:0018997). It has 15 children terms (in the link, click the neighbors title in the left side bar). 
 
```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "categories":["biolink:ChemicalEntity"]
                },
                "n1": {
                    "ids":["MONDO:0018997"],
                    "categories":["biolink:DiseaseOrPhenotypicFeature"],
                    "name": "noonan"
               }
            },
            "edges": {
                "eA": {
                    "subject": "n0",
                    "object": "n1",
                    "predicates": ["biolink:treats"]
                }
            }
        }
    }
}
```

The current PRs manipulate the QGraph so that the QNode has all 16 IDs for Noonan syndrome and its children. 

A result with one of the children (`MONDO:0008104` aka `Noonan syndrome 1`) is this one (remember `Noonan syndrome`'s ID is `MONDO:0018997`):
```
            {
                "node_bindings": {
                    "n0": [
                        {
                            "id": "PUBCHEM.COMPOUND:2907"
                        }
                    ],
                    "n1": [
                        {
                            "id": "MONDO:0008104"
                        }
                    ]
                },
                "edge_bindings": {
                    "eA": [
                        {
                            "id": "e505e0466360f09de4bb3415a2b83e85"
                        }
                    ]
                },
                "score": 0
            },
```

If my interpretation was followed, the QGraph wouldn't be manipulated (so the QNode would only have 1 ID, for Noonan syndrome specifically). Then the result above would use `query_id` and would look like this (remember `Noonan syndrome`'s ID is `MONDO:0018997`; the child term is `MONDO:0008104` aka `Noonan syndrome 1`)): 
```
            {
                "node_bindings": {
                    "n0": [
                        {
                            "id": "PUBCHEM.COMPOUND:2907"
                        }
                    ],
                    "n1": [
                        {
                            "id": "MONDO:0008104",
                            "query_id": "MONDO:0018997"
                        }
                    ]
                },
                "edge_bindings": {
                    "eA": [
                        {
                            "id": "e505e0466360f09de4bb3415a2b83e85"
                        }
                    ]
                },
                "score": 0
            },
```

</details>

### Concerns

I'm not sure if this is easy or not (I'm concerned about how the results-assembly will work if the QNode ID doesn't match IDs in the record). I'm also not sure if there will always be a 1-to-1 mapping of descendant ID to original QNode ID (what if a QNode had multiple IDs originally and 1 of them was a child of the other - which should be used as the `query_id`?), and if that'll cause any issues. 

An idea I have is this (it's great if we can think of something better):
* keep manipulating the QGraph/QNode by adding the descendant IDs, but somehow store which QNodes were expanded and the mappings between the original QNode IDs and its descendant IDs (this would allow us to find the original QNode ID when there are multiple to keep track of, see part C). 
* during results-assembly, for each result, look at the NodeBindings for the QNodes that were expanded. Iterate through each entry (there can be multiple entries when `is_set` was true), comparing the value of the `id` field to the stored original QNode IDs . If there are no exact matches, we can assume that the ID is a descendant -> then find which original QNode ID(s) it corresponds to, and add a `query_id` field with that original QNode ID to that NodeBinding. [Note](https://github.com/NCATSTranslator/ReasonerAPI/blob/master/TranslatorReasonerAPI.yaml#L477) that `id` and `query_id` fields are supposed to have 1 CURIE (string) as a value (not an array).  

The implications of introducing query_id were previously discussed [here](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/450#issuecomment-1175840395) (under Part B), and some of it is out-of-scope of this issue (how we ingest other TRAPI KPs that use it). EDIT: I don't think we need to worry about this if we kept something similar to the current PRs 2022/09/22 2:00 PM PT. See Notes section of https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1255488128 

---

## C. When there are multiple IDs to expand...

Can BTE handle cases where multiple QNodes have IDs that need expanding (and those are different namespaces)? What about 1 QNode that has multiple IDs but they're from multiple namespaces? 

This is more of a prompt for my manual testing (and maybe writing tests). 

<details>
<summary>examples</summary>

Explain Query: both "end" QNodes have an ID that could be expanded (different namespaces)
```
{
    "message": {
        "query_graph": {
            "edges": {
                "e00": {
                    "subject": "n0",
                    "object": "n1"
                },
                "e01": {
                    "subject": "n1",
                    "object": "n2"
                }
            },
            "nodes": {
                "n0": {                   
                    "ids": ["MONDO:0018997"],
                    "categories": ["biolink:Disease"],
                    "name": "noonan"
                },
                "n1": {
                    "categories": ["biolink:BiologicalProcessOrActivity"]
                },
                "n2": {
                    "ids": ["HP:0002315"],
                    "categories": ["biolink:PhenotypicFeature"],
                    "name": "headache"
                }
            }
        }
    }
}
```

EDIT: so there are two diff IDs that work for expansion
1 QNode with multiple IDs that could be expanded (different namespaces), and `is_set` is true (so maybe some results include stuff from both IDs). 
```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "categories":["biolink:ChemicalEntity"]
                },
                "n1": {
                    "ids":["MONDO:0018997", "HP:0002315"],
                    "categories":["biolink:DiseaseOrPhenotypicFeature"],
                    "is_set": true,
                    "name": "noonan and headache"
               }
            },
            "edges": {
                "eA": {
                    "subject": "n0",
                    "object": "n1",
                    "predicates": ["biolink:treats"]
                }
            }
        }
    }
}
```
</details>

**Comment by colleenXu** [comment]
And another note (D?): removing subclass / superclass relationships from x-bte annotation for the [ontologies](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1235045027) that node-expansion would cover...

EDIT: 9/7 group meeting notes: yes, we probably want to do that. But wait until after the ID stuff is implemented. Then one would do this

EDIT 9/22: on-hold because of the decision made for Part A. See edit in that section of original post

**Comment by colleenXu** [comment]
Su / Colleen Discussion 9/6 for Part A:

We then decided to go with Option 1:  @ericz1803. maybe disable ID expansion on creative-mode for now. He'll modify https://github.com/biothings/bte_trapi_query_graph_handler/pull/113
* Option 2: outer loop is ID expansion, inner loop is template library (but this gets into questions of template library design)
* Option 3: outer loop is template library, inner loop is ID expansion.  (NOTE, this design can leverage batch querying of IDs). Basically fits what I originally proposed. 

<details>
<summary>CX notes</summary>

* What is the desire (do we want to hit less templates with all expanded IDs, or hit more templates with the original ID first?)? related to template and library design...
    * brings up the idea of looping through expanded IDs vs looping through templates
* Option 2: an outer loop of iterating through original ID + descendants (inner loop is taking an id and going through all templates) (aka go through templates with original ID first, then try running each descendant disease ID one-by-one through all the templates)
* Option 3: go through all the templates once, with the expanded IDs (original ID + descendants). Basically fits what I envisioned...
</details>

**Comment by colleenXu** [comment]
From 9/7 group meeting:

Part B: pause for discussion with the rest of Translator. I'll make some drafts of what we think the intended behavior is (with notes), to share with Translator. 

Part C: @ericz1803 has confirmed that the code should work in the examples given.

**Comment by colleenXu** [comment]
## Bringing up issues around Part B

We're confused when we compare what's in the Architecture document to what seems to be the intent of TRAPI 1.3's NodeBinding.query_id. It's not clear what should be done to the query_graph, knowledge_graph, and results sections when ID/QNode-expansion is done. 

<details>
<summary>I'll use a one-hop Predict-style as an example here.</summary>

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "categories":["biolink:ChemicalEntity"]
                },
                "n1": {
                    "ids":["MONDO:0018997"],
                    "categories":["biolink:DiseaseOrPhenotypicFeature"],
                    "name": "noonan syndrome"
               }
            },
            "edges": {
                "eA": {
                    "subject": "n0",
                    "object": "n1",
                    "predicates": ["biolink:treats"]
                }
            }
        }
    }
}
```

</details>

### query_graph

It sounds like when ID-expansion is done, the query-graph should not be mutated aka the ARA/KP shouldn't add additional IDs to the original QNodes. Is that right?

### knowledge_graph and results

In the Architecture [document](https://github.com/NCATSTranslator/TranslatorArchitecture/blob/master/README.md), it says:
> KPs that implement the Translator Reasoner API must perform the following kinds of reasoning in answering queries:
>     Making identifiers more specific, e.g. responding to a query involving an entity with information related to a subclass of that entity. In the knowledge_graph portion of the response, the more-specific identifier must be present and linked to the less-specific identifier. In the results portion of the response, the more-specific response node will be bound to the less-specific query node.

However, this seems a bit odd to me. Looking at the one-hop Predict query, what if the original (less-specific ID) doesn't have direct connections to any "answers"? Does it still need to be in the KG? 

And if it is in the KG, does the original need to be connected to the more-specific ID entities with "subclass_of" Edges (with the right directionality)? And then does the original Node and the "subclass_of" Edge need to be included in the result somehow? 

My opinion is in the case of the one-hop Predict query, the original (less-specific) doesn't need to be there when it doesn't have any direct connections to the "answers" and no "subclass_of" Edges are needed. I think all of that info is in the TRAPI 1.3 NodeBinding.query_id vs id fields, and that ideally the UI would process those fields and present the result in a way that makes it clear that the more-specific ID is related to the original (less-specific) ID used in the QNode.  

### To illustrate this...

Imagine the one-hop Predict query above had 1 "answer" Edge, between a child of Noonan syndrome and a chemical. AKA there were no Edges found between Noonan syndrome itself (MONDO:0018997) and a chemical. 

<details>
<summary>simplified TRAPI response with no Noonan syndrome node (uses NodeBinding.query_id)</summary>

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "categories": ["biolink:ChemicalEntity"]
                },
                "n1": {
                    "ids": ["MONDO:0018997"],
                    "categories": ["biolink:DiseaseOrPhenotypicFeature"],
                    "name": "noonan syndrome"
                }
            },
            "edges": {
                "eA": {
                    "subject": "n0",
                    "object": "n1",
                    "predicates": ["biolink:treats"]
                }
            }
        },
        "knowledge_graph": {
            "nodes": {
                "PUBCHEM.COMPOUND:2907": {
                    "categories": ["biolink:SmallMolecule"],
                    "name": "Cyclophosphamide",
                    "attributes": []
                },
                "MONDO:0008104": {
                    "categories": ["biolink:Disease"],
                    "name": "Noonan syndrome 1",
                    "attributes": []
                }
            },
            "edges": {
                "e505e0466360f09de4bb3415a2b83e85": {
                    "predicate": "biolink:treats",
                    "subject": "PUBCHEM.COMPOUND:2907",
                    "object": "MONDO:0008104",
                    "attributes": []
                }
            }
        },
        "results": [
            {
                "node_bindings": {
                    "n0": [
                        {
                            "id": "PUBCHEM.COMPOUND:2907"
                        }
                    ],
                    "n1": [
                        {
                            "id": "MONDO:0008104",
                            "query_id": "MONDO:0018997"
                        }
                    ]
                },
                "edge_bindings": {
                    "eA": [
                        {
                            "id": "e505e0466360f09de4bb3415a2b83e85"
                        }
                    ]
                },
                "score": 0
            }
        ]
    }
}
```

</details>

VS adding the Noonan syndrome node with an "subclass_of" edge. I used the NodeBinding.query_id format for the result, but then it's awkward that the original "noonan syndrome" node and the "subclass_of" edge is in the KG section but not part of the result. 

<details>
<summary>adding the Noonan syndrome node with an "subclass_of" edge</summary>

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "categories": ["biolink:ChemicalEntity"]
                },
                "n1": {
                    "ids": ["MONDO:0018997"],
                    "categories": ["biolink:DiseaseOrPhenotypicFeature"],
                    "name": "noonan syndrome"
                }
            },
            "edges": {
                "eA": {
                    "subject": "n0",
                    "object": "n1",
                    "predicates": ["biolink:treats"]
                }
            }
        },
        "knowledge_graph": {
            "nodes": {
                "PUBCHEM.COMPOUND:2907": {
                    "categories": ["biolink:SmallMolecule"],
                    "name": "Cyclophosphamide",
                    "attributes": []
                },
                "MONDO:0008104": {
                    "categories": ["biolink:Disease"],
                    "name": "Noonan syndrome 1",
                    "attributes": []
                },
                "MONDO:0018997": {
                    "categories": ["biolink:Disease"],
                    "name": "Noonan syndrome",
                    "attributes": []
                }
            },
            "edges": {
                "e505e0466360f09de4bb3415a2b83e85": {
                    "predicate": "biolink:treats",
                    "subject": "PUBCHEM.COMPOUND:2907",
                    "object": "MONDO:0008104",
                    "attributes": []
                },
                "subclass_edge_1": {
                    "predicate": "biolink:subclass_of",
                    "subject": "MONDO:0008104",
                    "object": "MONDO:0018997",
                    "attributes": []
                }
            }
        },
        "results": [
            {
                "node_bindings": {
                    "n0": [
                        {
                            "id": "PUBCHEM.COMPOUND:2907"
                        }
                    ],
                    "n1": [
                        {
                            "id": "MONDO:0008104",
                            "query_id": "MONDO:0018997"
                        }
                    ]
                },
                "edge_bindings": {
                    "eA": [
                        {
                            "id": "e505e0466360f09de4bb3415a2b83e85"
                        }
                    ]
                },
                "score": 0
            }
        ]
    }
}
```

</details>

**Comment by colleenXu** [comment]
Decision from 9/19 Su and me meeting:
* Part A: @ericz1803 should make the changes needed to get creative-mode working. We [decided](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/497#issuecomment-1238482553) to try disabling ID/QNode expansion for creative-mode rather than [other](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/497#issue-1359693815) [options](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/497#issuecomment-1238482553). 
    * we can't merge the PRs to main before this is addressed
* Part B: There are still discussions happening in the Translator consortium. We can move forward with these PRs even though we're not sure of the response "format". The current PRs don't use `query_id` in the results or include "subclass_of" edges between the original QNode-ID-entities and descendant-ID-entities.

I'm doing some review of the PRs while keeping this stuff about Part B in mind (and not testing creative-mode directly because Part A's stuff hasn't been done yet). See posts like https://github.com/biothings/BioThings_Explorer_TRAPI/issues/366#issuecomment-1255506397, https://github.com/biothings/BioThings_Explorer_TRAPI/issues/196#issuecomment-1255530559, and https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1255488128

**Comment by colleenXu** [comment]
Also noting: I interpret the Architecture principle on ID/QNode-expansion [above](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/497#issuecomment-1242482638) as saying original QNode IDs should be expanded...and it's maybe okay if IDs used as input for intermediate hops aren't expanded. 

The opening post of this PR says that it works on the original QNode IDs given https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issue-1313650345 and this issue of "what about intermediate hops" was discussed in [these](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1235002384) [later](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1235003011) [posts](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1235049161)

**Comment by colleenXu** [comment]
Currently (2022-10-16), the linked [query-handler PR](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113) will run Creative-mode because initially no ID-expansion happens. However, as the templates run, **ID-expansion is done**. Andrew and I discussed this 2022-10-17 and decided that this behavior was fine.


Also noted in the second point in the "Notes" section of this [post](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1280192697). 

---

However, when these PRs are pushed:
* some [templates](https://github.com/biothings/bte_trapi_query_graph_handler/tree/main/data/templates/Drug-treats-Disease) will need to be adjusted to remove the subclass_of hops.
* we may want to remove all x-bte annotations that have subclass_of / superclass_of predicates.

**Comment by colleenXu** [comment]
Andrew and I agreed that this is a necessary addition for this first iteration of ID/node-expansion:

When ID-expansion is done on a QNode, check that this QNode has the property `is_set: true` set. If the QNode doesn't have this property, add it. 

For example:

<details>

<summary>original query</summary>

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "ids": ["MONDO:0005301"],
                    "categories": ["biolink:Disease"],
                    "name": "multiple sclerosis",
                    "is_set": true
                },
                "n1": {
                    "categories":["biolink:Gene"]
                },
                "n2": {
                    "ids":["GO:0042552"],
                    "categories":["biolink:BiologicalProcess"],
                    "name": "myelination"
               }
            },
            "edges": {
                "e0": {
                    "subject": "n0",
                    "object": "n1"
                },
                "e1": {
                    "subject": "n2",
                    "object": "n1"
                }
            }
        }
    }
}
```

</details>

...would have ID-expansion done on QNodes n0 and n2. We want the `is_set: true` property added to n2 (n0 already has it). 

<details>

<summary>modified query with ID-expansion and is_set</summary>

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "ids": [
                        "MONDO:0005301", "MONDO:0005284", "MONDO:0005314", 
                        "MONDO:0012956", "MONDO:0012957", "MONDO:0012958",
                        "MONDO:0000450", "MONDO:0000451", "MONDO:0000452"
                        ],
                    "categories": ["biolink:Disease"],
                    "name": "multiple sclerosis",
                    "is_set": true
                },
                "n1": {
                    "categories":["biolink:Gene"]
                },
                "n2": {
                    "ids":[
                        "GO:0042552", "GO:0022010", "GO:0022011"
                        ],
                    "categories":["biolink:BiologicalProcess"],
                    "name": "myelination",
                    "is_set": true
               }
            },
            "edges": {
                "e0": {
                    "subject": "n0",
                    "object": "n1"
                },
                "e1": {
                    "subject": "n2",
                    "object": "n1"
                }
            }
        }
    }
}
```
</details>

---

Second, this is not necessary but would be nice to have: TRAPI-response logs (INFO level) for ID-expansion and adding `is_set: true` to QNodes. 

Something similar to existing console logs:

```
  bte:biothings-explorer-trapi:main Expanded ids for node n0: (1 ids -> 9 ids) +0ms
  bte:biothings-explorer-trapi:main Expanded ids for node n2: (1 ids -> 3 ids) +0ms
```

<details>

<summary>Like this</summary>

```
        {
            "timestamp": "2022-11-04T01:49:03.605Z",
            "level": "INFO",
            "message": "Expanded ids for node n0: (1 ids -> 9 ids)",
            "code": null
        },
        {
            "timestamp": "2022-11-04T01:49:03.605Z",
            "level": "INFO",
            "message": "Expanded ids for node n2: (1 ids -> 3 ids)",
            "code": null
        },
        {
            "timestamp": "2022-11-04T01:49:03.605Z",
            "level": "INFO",
            "message": "Added is_set: true to node n2",
            "code": null
        },
```
</details>

---

Third, we have a question for @ericz1803:  

I notice that the console-logs for ID expansion occur very early, before logs like `Start to load metakg` log or `(1) Creating edges for manager`. 

Is that expected, and happening because ID-expansion is happening very early in the process? 

And if ID-expansion is happening that early, does it make sense to make it later in the process (like after loading the metaKG, but before creating edges for the edge-manager)?

(Noting: organized points originally in this [post](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1280213218) and this [post](https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1299583992))

**Comment by colleenXu** [comment]
Noting multiple PRs / branches related to this issue:

These two go together:
* https://github.com/biothings/bte-trapi-workspace/pull/17
* https://github.com/biothings/bte_trapi_query_graph_handler/pull/113
    * I recently added a commit to adjust templates - removing the subclass operations and rearranging the order https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1304412147

---

We could also make PRs from these branches and merge them. They remove subclass/superclass x-bte operations ([mentioned](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/497#issuecomment-1281419468) in the comments of this post). ~It's not clear that we need them when we automatically do ID-expansion...~
* https://github.com/NCATS-Tangerine/translator-api-registry/tree/remove-subclassing
* https://github.com/biothings/BioThings_Explorer_TRAPI/tree/remove-subclassing

EDIT: However, the Automat APIs would still have superclass/subclass operations. Hmmm....but it could still help to remove these operations. 

Taking a bigger picture, what we want to do is avoid redundant behavior (doing automatic ID-expansion on a QNode ID and then traversing a hop that also does a "subclass_of" operation). 

Another way to do this: if a QGraph has the predicates "subclass_of" or "superclass_of" on a QEdge, don't execute the query. Return an error message saying `BTE automatically 'expands' IDs to include ontological-descendants for the following ontologies: X, Y, Z. BTE therefore does not accept QGraphs with the predicates "subclass_of" or "superclass_of" on a QEdge`.

**Comment by ericz1803** [comment]
I pushed 2 changes that set the QNode `is_set` property to `true` for expanded nodes and logs it.

To your question, the ID expansion is happening when the QueryGraph is set, so very early in the process. I think it makes sense to do it here because the ID expansion doesn't depend on metaKG being loaded and it is the most straightforward place to do it.

**Comment by tokebe** [comment]
@colleenXu @andrewsu with node-expansion having been deployed to Prod, can we close this and the following related issues?
- #196 
- #366

**Comment by colleenXu** [comment]
oh yay! okay, going to close the issues.

**Comment by colleenXu** [comment]
Noting the ontologies/vocabs covered, for easy ref. Got from my old post https://github.com/biothings/bte_trapi_query_graph_handler/pull/113#issuecomment-1280192697)
* MONDO
* HP
* GO (BP, MF, CC)
* DOID
* CHEBI
* UMLS
